design/ratelimit-design.md

```diff
@@ -1,175 +1,502 @@
-Contour will support rate limiting via the `remote_address` presented to Envoy generically as well as allowing users to specify an IP address which will allow for blocking specific IPs (by defining a `requests_per_unit` of zero).
+For global rate limiting, the user defines the descriptors to be generated and sent to the external RLS.
+Descriptors contain entries including things like: the client IP, the value of a particular header, the destination cluster, etc.
+The external RLS makes the rate limit decision based on the descriptors, and returns either a 200 or a 429 to Envoy.
+The operator of the external RLS must configure it with actual rate limits for different descriptors.
+Each RLS implementation may have its own configuration format and mechanism for defining rate limits for descriptors, so Contour cannot provide a generic API for defining these.
 
-In the event a user defines `rateLimits` via IngressRoute, but are not enabled, the status field of that object will be updated to reflect the error.
-If an Ingress object has the appropriate annotation, a log message will be written which will explain the error encountered.
+
+For global rate limiting, an `ExtensionService` is defined to map to an external Rate Limit Service.
+The RLS is defined in the Contour config file, which will be used by any `HTTPProxies` that define a global rate limit policy.
 
 ## Detailed Design
 
-### Lyft RateLimit Implementation
+### RateLimitPolicy type
+
+A sample `RateLimitPolicy` looks like:
+
+```yaml
+rateLimitPolicy:
+  # local defines local rate limiting properties for the virtual host or route.
+  local:
+    # requests defines how many requests per unit of time to allow.
+    # This programs the "tokens_per_fill" field on the Envoy local
+    # rate limit filter.
+    # See ref. https://www.envoyproxy.io/docs/envoy/latest/api-v3/type/v3/token_bucket.proto#envoy-v3-api-msg-type-v3-tokenbucket.
+    requests: 100
+    # unit defines the period of time within which requests over the
+    # limit will be rate limited. This programs the "fill_interval" field
+    # on the Envoy local rate limiter.
+    # See ref. https://www.envoyproxy.io/docs/envoy/latest/api-v3/type/v3/token_bucket.proto#envoy-v3-api-msg-type-v3-tokenbucket.
+    unit: second
+    # burst defines how many additional requests above the baseline requests
+    # are allowed in a short period of time. This, along with "requests", 
+    # programs the "max_tokens" field on the Envoy local rate limit filter.
+    # See ref. https://www.envoyproxy.io/docs/envoy/latest/api-v3/type/v3/token_bucket.proto#envoy-v3-api-msg-type-v3-tokenbucket.
+    burst: 20
+  # global defines global rate limiting properties for the virtual host or route.
+  global:
+    # rateLimitDescriptors defines the descriptors to be generated and sent to the
+    # external Rate Limiting Service (RLS) for a rate limit decision.
+    rateLimitDescriptors:
```

Comment:
Should `rateLimitDescriptors` just be named `descriptors` since we're already inside a `rateLimitPolicy`?

Reply:
+1 from me, will update.